### DISCUSSION TITLE
Overlapping Edges Obscures Meaning of Tree

### DISCUSSION THREAD

**Discussion by SleighJ** [xyflow/xyflow#2757] [Layouting]
Hey All,

Loving the library and the work all of you have put into it.

Reaching out because it seems that I have either hit an edge case or made a mistake in my implementation and I'm unable to fix it (though the documentation is quite thorough 👏 ). Hoping I have just missed something along the way.

I have the following spec - it's edges routed/spaced out so the node's sources and targets are recognizable and one can deduce meaning from it's placement in the tree
![Screenshot 2023-01-17 at 5 12 34 PM](https://user-images.githubusercontent.com/20913377/213044571-ecbff57a-c61b-410f-b037-3128f0b58c4b.png)

After coding the spec, the issue below happened with multiple different trees. Notice the 'Action' node (at the very bottom of the 3rd layer) is easily assumed to be a 'Field' node as the edges pertaining to each node type overlap - this happens even though they extend from different `<Handles />`.
![Screenshot 2023-01-17 at 5 23 14 PM](https://user-images.githubusercontent.com/20913377/213046796-377dc9cd-1a8f-4931-a783-2f35ae7d2c3b.png)

My issue occurs with both the Dagre and Elk layout libraries. After several unsuccessful implementations, I recreated the tree using Elk.js' sandbox (which seems to be down right now) ([ELK Editor](https://rtsys.informatik.uni-kiel.de/elklive/index.html)) and was able to pass the necessary options to get the tree to render without overlapping edges - unfortunately, those options are either overwritten or obscured by the specs use of the handles (1-to-many relationships and multiple handles per node). Any insight on this would be much appreciated.

I've tried grouping nodes via the `parentNode` prop, but haven't had any luck since only a parent node's id is reference-able (and not the id of the `<Handle />` it stems from).
 
Wondering if there is anything I can do to help my case short of changing the edge type to the default (bezier) or requesting other changes to the design.

**Comment by moklick** [comment]
Hey @SleighJ 

we don't have smart edge routing. Your screenshot shows the expected behaviour. Unfortunately there is no easy way to fix this with the help of the library. You could implement a custom edge with help of elks edge routing. 

Maybe this helps: https://github.com/tisoap/react-flow-smart-edge

**Reply by SleighJ** [reply]
Thanks for the quick reply @moklick 

I've tried react-flow-smart-edge, unfortunately the step-style edge wasn't up to par (rendered diagonals, flashed on some re-renders etc). Also, there didn't seem to be any behavior preventing overlap of edges, just the intersections of edges and nodes (though I could be wrong here, I didn't spend quite as much time with this package - will revisit today).

Curious as to what your approach would be with a custom edge and the elks edge routing. How would I go about making this different than what comes out of the box with RF?

As far as I could tell, the custom edge was mostly meant for changes in styling as the functionality was more or less under the hood of RF. Is there a way to bypass that and use the routing from Elk by chance? I was able to retrieve the intersecting SVG edge paths, but redrawing the edge SVGs seemed like a big bit of logic, and I'm not sure those efforts will yield less edge cases than what I'm currently experiencing (these trees will be created dynamically, solving for one particular case won't do me much good here).

**Reply by kvz** [reply]
> Maybe this helps: https://github.com/tisoap/react-flow-smart-edge

react-flow-smart-edge is compatible with 'reactflow', but not with '@xyflow/react'. I tried and hit the same as this locked issue here: https://github.com/tisoap/react-flow-smart-edge/issues/61, doesn't look like they will fix this. Meanwhile it's the first thing I hit trying to make a diagram. When two components interact at multiple stages in a process, the labels will stack and you can't see them, without implementing custom collision detection logic.

**Reply by ArjixWasTaken** [reply]
Sorry for necrobump: it does support `@xyflow/react` but it became deprecated since the 17th

**Reply by tisoap** [reply]
I've updated [react-flow-smart-edge](https://github.com/tisoap/react-flow-smart-edge) to work with React Flow 12 (`@xyflow/react` package)

**Comment by sudarsanarao995** [comment]
@moklick , any solution for that or is it possible overlapping Edges intersection point will make some circular or broken design ?

**Reply by aziz-haddad** [reply]
I finally managed to fix the issue with the overlapping edges by adding multiple handles to each node so the edges connect to different handles and thus don't overlap. Check the docs for multiple handles here. https://reactflow.dev/learn/customization/handles

**Comment by awaisshah228** [comment]
I built an open-source package that solves this: **[avoid-nodes-edge](https://www.npmjs.com/package/avoid-nodes-edge)**

Edges automatically route around nodes with clean orthogonal paths. Uses libavoid (C++ WASM) in a Web Worker for performance. Parallel edges are nudged apart so they never overlap.

[Live Demo](https://avoid-nodes-pro-example.vercel.app) | [npm](https://www.npmjs.com/package/avoid-nodes-edge) | [GitHub](https://github.com/awaisshah228/react-flow-avoid-nodes-routing) | [Package source](https://github.com/awaisshah228/react-flow-avoid-nodes-routing/tree/turbo-package)

`npm install avoid-nodes-edge l`

Works with React Flow v12+.

**Reply by pinesh** [reply]
Really great!

**Comment by tisoap** [comment]
I'm the creator of [react-flow-smart-edge](https://github.com/tisoap/react-flow-smart-edge) , a library that uses using A* pathfinding to route edges around other nodes, avoiding intersections. It has smart equivalents for every type of edge React Flow has. It's been updated to work with React Flow 12.
